src/common/meta/src/key/tombstone.rs

```diff
@@ -0,0 +1,448 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use snafu::{ensure, OptionExt};
+
+use super::TableMetaKeyGetTxnOp;
+use crate::error::{self, Result};
+use crate::key::txn_helper::TxnOpGetResponseSet;
+use crate::kv_backend::txn::{Compare, CompareOp, Txn, TxnOp};
+use crate::kv_backend::KvBackendRef;
+
+/// [TombstoneManager] provides the ability to:
+/// - logically delete values
+/// - restore the deleted values
+pub(crate) struct TombstoneManager {
+    kv_backend: KvBackendRef,
+}
+
+const TOMBSTONE_PREFIX: &str = "__tombstone/";
+
+pub(crate) struct TombstoneKey<T>(T);
+
+fn to_tombstone(key: &[u8]) -> Vec<u8> {
+    [TOMBSTONE_PREFIX.as_bytes(), key].concat()
+}
+
+impl TombstoneKey<&Vec<u8>> {
+    /// Returns the origin key and tombstone key.
+    fn to_keys(&self) -> (Vec<u8>, Vec<u8>) {
+        let key = self.0;
+        let tombstone_key = to_tombstone(key);
+        (key.clone(), tombstone_key)
+    }
+
+    /// Returns the origin key and tombstone key.
+    fn into_keys(self) -> (Vec<u8>, Vec<u8>) {
+        self.to_keys()
+    }
+
+    /// Returns the tombstone key.
+    fn to_tombstone_key(&self) -> Vec<u8> {
+        let key = self.0;
+        to_tombstone(key)
+    }
+}
+
+impl TableMetaKeyGetTxnOp for TombstoneKey<&Vec<u8>> {
+    fn build_get_op(
+        &self,
+    ) -> (
+        TxnOp,
+        impl FnMut(&'_ mut TxnOpGetResponseSet) -> Option<Vec<u8>>,
+    ) {
+        let key = to_tombstone(self.0);
+        (TxnOp::Get(key.clone()), TxnOpGetResponseSet::filter(key))
+    }
+}
+
+/// Atomic Key:
+/// The value corresponding to the key remains consistent between two transactions.
+pub(crate) enum Key {
+    Atomic(Vec<u8>),
+    Other(Vec<u8>),
```

Comment:
I don't think "Other" is a good name for an enum variant. Maybe it's better just 
```rust
struct Key {
  bytes: Vec<u8>,
  atomic: bool,
}
```